slsDetectorSoftware/src/Module.cpp

```diff
@@ -2386,93 +2386,158 @@ uint64_t Module::getReceiverCurrentFrameIndex() const {
+    static_assert(!std::is_pointer<ARG>::value,                                \
+                  "Pointer type is incompatible with templated " DST);         \
+    static_assert(!std::is_same<ARG, std::nullptr_t>::value,                   \
+                  "nullptr_t type is incompatible with templated " DST);
+
 void Module::sendToDetector(int fnum, const void *args, size_t args_size,
-                            void *retval, size_t retval_size) {
+                            void *retval, size_t retval_size) const {
+    // This is the only function that actually sends data to the detector
+    // the other versions use templates to deduce sizes and create
+    // the return type
+    preSendArgsCheck(args, args_size, retval, retval_size);
     auto client = DetectorSocket(shm()->hostname, shm()->controlPort);
     client.sendCommandThenRead(fnum, args, args_size, retval, retval_size);
     client.close();
 }
 
+void Module::sendToDetector(int fnum, const void *args, size_t args_size,
+                            void *retval, size_t retval_size) {
+    static_cast<const Module &>(*this).sendToDetector(fnum, args, args_size,
+                                                      retval, retval_size);
+}
+
 template <typename Arg, typename Ret>
-void Module::sendToDetector(int fnum, const Arg &args, Ret &retval) {
+void Module::sendToDetector(int fnum, const Arg &args, Ret &retval) const {
     LOG(logDEBUG1) << "Sending: ["
                    << getFunctionNameFromEnum(static_cast<detFuncs>(fnum))
                    << ", nullptr, 0, " << typeid(Ret).name() << ", "
                    << sizeof(Ret) << "]";
+    STATIC_ASSERT_ARG(Arg, "sendToDetector")
+    STATIC_ASSERT_ARG(Ret, "sendToDetector")
     sendToDetector(fnum, &args, sizeof(args), &retval, sizeof(retval));
     LOG(logDEBUG1) << "Got back: " << ToString(retval);
 }
 
+template <typename Arg, typename Ret>
+void Module::sendToDetector(int fnum, const Arg &args, Ret &retval) {
+    static_cast<const Module &>(*this).sendToDetector(fnum, args, retval);
+}
+
 template <typename Arg>
-void Module::sendToDetector(int fnum, const Arg &args, std::nullptr_t) {
+void Module::sendToDetector(int fnum, const Arg &args, std::nullptr_t) const {
     LOG(logDEBUG1) << "Sending: ["
                    << getFunctionNameFromEnum(static_cast<detFuncs>(fnum))
                    << ", " << typeid(Arg).name() << ", " << sizeof(Arg)
                    << ", nullptr, 0 ]";
+    STATIC_ASSERT_ARG(Arg, "sendToDetector")
     sendToDetector(fnum, &args, sizeof(args), nullptr, 0);
 }
 
+template <typename Arg>
+void Module::sendToDetector(int fnum, const Arg &args, std::nullptr_t) {
+    static_cast<const Module &>(*this).sendToDetector(fnum, args, nullptr);
+}
+
 template <typename Ret>
-void Module::sendToDetector(int fnum, std::nullptr_t, Ret &retval) {
+void Module::sendToDetector(int fnum, std::nullptr_t, Ret &retval) const {
     LOG(logDEBUG1) << "Sending: ["
```

Comment:
Since now we have const sendToDetector, we can set many of the get Module class functions to a const..

Reply:
Indeed, I put this off before since I didn't like the code duplication. Should I add it in this PR?

Reply:
upto you